### DISCUSSION TITLE
New Feature - Z-Probe Offset Calibration

### DISCUSSION THREAD

**Discussion by huntervaneck** [supermerill/SuperSlicer#1722] [Community contribution]
Hello,

I'm proposing an addition to the Calibration menu called "Z-Probe Offset Calibration" that would generate g-code files for setting the proper offset between a BLtouch (for example) and the nozzle. The simplest way to do this in my opinion would be to generate a file similar to the flow calibration tool, Which would give you a coarse adjustment and a fine adjustment with z-offset values ranging from negative to positive. 

"Z-Probe Offset Calibration" should require Bed leveling first and then Flow Rate calibration before being implemented. The reason I believe the flow rate should come before this calibration is that an improper flow rate calibration would potentially represent an improper line width for the first layer and give inaccurate results when trying to set the proper z-probe offset.

Currently, I do this process while the machine is printing. I will Baby-step Z until I'm close to the first layer finish that I want. This isn't ideal and a simple calibration tool with number values to represent offset would be very helpful.

Thank you,
Hunter

**Comment by supermerill** [comment]
How is this different from bel level calibration then flow rate calibration?

And the flow rate test depends on the bed level, but the opposite is less prevalent, as the flow rate is often in the good ballpark, and it won't impact the bed test unless there is something massively wrong.

**Comment by huntervaneck** [comment]
Bed leveling calibration helps to set the distance between the nozzle to the print bed so that all four corners are equidistant from the nozzle and "level"

Flow rate helps to set the line width so that when the software says it's printing .48mm, you can accurately measure .48mm with calipers.

Z probe offset would be for when you've got the bed level and the flow rate set,  but the spacing between your line widths is off. This means that the BLtouch offset to the nozzle is not set correctly. This could happen after replacing the BLtouch probe tip after it breaks or exchanging nozzles but has nothing to do with the bed not being level or the flow rate being wrong. It would move the bed closer or further away to fill the spacing between each line. 

This feature would become incredibly helpful for myself and other people with multiple printers that are all the same. Each printer is leveled and their flow rates are all the same, however, their z-probe offsets are all different, due to different tolerances within parts. Having a simple calibration tool that creates a stair step, starting with a negative number and moving up would save a ton of time because I could run it on every machine and fine-tune the offset to close the spacing between the lines.

I hope this makes sense, it's somewhat hard to explain over text.

**Reply by supermerill** [reply]
> Bed leveling calibration helps to set the distance between the nozzle to the print bed so that all four corners are equidistant from the nozzle and "level"

Current bed calibration ensure that you are at the right level from the get-go, by looking at the patch.
If you want to be extra precise, you can do bed-level -> flow -> bed-level
The calibration is mostly useful for beds that are levelled by screws.

> Flow rate helps to set the line width so that when the software says it's printing .48mm, you can accurately measure .48mm with calipers.

No, it's not exactly how it works.
We hope it's 0.48 but depending of the previous path, temperature and filament type, it may not. And there is nothing to change the theoretical extruded filament geometry.
The calibration ensures that when you want to extrude 1mm3, it extrudes 1mm3.

> Each printer is leveled and their flow rates are all the same, however, their z-probe offsets are all different, due to different tolerances within parts.

I can conceive that printing 5 patches is too much for a bt-touch printer, as it only really need one. I can modify the calibration to set an option for "one patch only - multiple heights", that is what you ask for?

**Reply by huntervaneck** [reply]
> No, it's not exactly how it works.
We hope it's 0.48 but depending of the previous path, temperature and filament type, it may not. And there is nothing to change the theoretical extruded filament geometry.
The calibration ensures that when you want to extrude 1mm3, it extrudes 1mm3.

Okay, I understand now - That makes sense to me.

> I can conceive that printing 5 patches is too much for a bt-touch printer, as it only really need one. I can modify the calibration to set an option for "one patch only - multiple heights", that is what you ask for?

This would be perfect for what I'm trying to do. It would allow me to swap out hardware and have all settings stay the same, except for the probe offset to the bed. This way you could visually see which height is the best and then apply it to the current offset. Typically these offsets that I'm referring to are small, such as when changing a nozzle it can offset the probe by -0.1mm or smaller.

**Reply by supermerill** [reply]
https://github.com/supermerill/SuperSlicer/issues/1725

**Comment by huntervaneck** [comment]
For example, a calibration tool to set my auto bed leveling probe offset would allow me to change the nozzles on all 10 of my printers, swap out build plates, etc. and instead of running the bed leveling calibration (since it's got a BLtouch) or the flow rate calibration (since I already know that value), I would simply run Z-probe offset calibration and use that as a gauge to adjust by -0.50mm to -0.100mm to close the gap between the printed lines.

**Reply by freighttt** [reply]
is this not the very reason we calibrate the z offset and then save it on the printer side? Im not sure I understand what problem this is trying to solve. How often are you changing to unknown nozzles/probes resulting in an offset change greater than the "margin of error"?

Calibrate each machine. Hardcode the most used z offset in the firmware. When you change nozzles or plates or materials(that need more or less squish), the distance changes a known value, unless its new or the machine isnt stiff enough for repeatability. When you change them, but you dont want it to be permanently changed in the firmware, add the needed change to the start gcode. Keep a log of this. I have a small list beside my machine specifying the needed adjustments between different nozzles and beds. Its easy enough to create new machine profiles to handle this as well. 

Do we really need a calibration test/function to print a test patch instead of watching and adjusting the first layer as it goes down? Running a calibration to find that value is going to take longer than simply adjusting the first print on the fly. Use a few extra skirt lines if you want enough time to be extra sure before it starts the object. Even with a calibration test, printing a single patch is not going to give you anything more than a ballpark value that will still almost certainly need visual adjustment.

**Reply by huntervaneck** [reply]
> is this not the very reason we calibrate the z offset and then save it on the printer side?

When calibrating the z offset on the printer side, it's still not a perfect process. Let's say that the heated bed is warmed up, the BLtouch probe makes contact with the center of the bed and then you lower the nozzle to the bed to set the offset with a piece of paper. Due to the convex curve in the glass bed introduced by the heat, and the X offset of the nozzle with respect to the probe, this would then cause the nozzle to introduce an unknown error in the Z-axis. The error is small, but not insignificant in terms of the spacing it introduces between line widths while printing.

> Do we really need a calibration test/function to print a test patch instead of watching and adjusting the first layer as it goes down?

No, we don't need a calibration, it would just be very helpful. Watching and adjusting a skirt as it's being printed isn't as intuitive as it sounds, considering your view is usually blocked by fans or shrouds. Maybe it's just me, but I also don't always have the time to sit and watch a print to adjust the height. I run 10 printers and so that's 10x the amount of time to adjust the height. 

> printing a single patch is not going to give you anything more than a ballpark value that will still almost certainly need visual adjustment.

I agree with you, but that's also not what I meant to be generated. A single flat patch wouldn't do me any good. I was saying that if a file was generated that had multiple stair steps in it with numerical values to show you what height the steps were printing at, you could then use that to adjust for a height with no gaps between print lines. I could simply run the calibration on every printer and come back when it's done. This would give me a value that I could quickly run through and add or subtract from the Z-Probe offset value. 

( example image not to scale)

![Part1](https://user-images.githubusercontent.com/30808296/139279537-db8960aa-5ac6-4bb6-ba81-9bb3dd311d1a.JPG)

If this is too much work or you don't find it to be a valuable addition nor a priority, I completely understand. But I could see myself using it often to maintain printers and also to help with perfecting IDEX height adjustment, which is currently a very rudimentary process. Like you mentioned before, there are 10 ways to do everything, but having the calibration built into the slicer where it auto-generates the necessary files and explains the process is incredibly helpful, as setting up these files is not always very straightforward. 

Even if you forget everything I said above and just add a calibration tab for IDEX X, Y and Z offset alignment, that would be amazing.

**Reply by freighttt** [reply]
> > is this not the very reason we calibrate the z offset and then save it on the printer side?
> 
> When calibrating the z offset on the printer side, it's still not a perfect process. Let's say that the heated bed is warmed up, the BLtouch probe makes contact with the center of the bed and then you lower the nozzle to the bed to set the offset with a piece of paper. Due to the convex curve in the glass bed introduced by the heat, and the X offset of the nozzle with respect to the probe, this would then cause the nozzle to introduce an unknown error in the Z-axis. The error is small, but not insignificant in terms of the spacing it introduces between line widths while printing.
> 

This is why you specify the probe offset X and Y in the firmware. You should be building a mesh, using the visualizer to get the bed as flat as possible. This process is not intended to be the precise way of getting your nozzle offset, but for building the bed mesh. Once your mesh is built, and the bed is flat relative to the probe, you adjust your z offset while printing. There is no discrepancy between the location of the nozzle and the location of the probe, because the mesh is placed relative to your end stops. The X and Y dont change, so the mesh is always in the same place. Where your z endstop(presumably your probe) triggers during the homing process is irrelevant to the mesh except in the Z. The z offset you choose is disrupted the the probe/nozzle offset only if you were to place the probe in a different location on the bed every time you home. Some people go so far as to only use their probe for mesh building, and use a separate Z end stop for homing. There are various reasons why they do this, but it illustrates the point that probe to nozzle offset is irrelevant. With properly configured firmware, there is no discrepancy, since the nozzle and probe are operating over the same absolutely positioned mesh.  If you tell the machine to home, then probe at 150/150, and then send the nozzle to 150/150, they should both go to the same place. Assuming you are homing in the same location every single time, and are using the mesh built for the temperature assigned(you do have a mesh for each bed temp, right?), 0/0/0 should put the tip of the nozzle in the exact same spot every time. The only reason that should change is a change in nozzle height. Even then, that should be not more than a couple clicks on the baby stepping at the beginning of the print. To design and implement a new calibration test to aid something that already has a more effectual feature(baby stepping, available on all firmware) designed to address it, seems like a strange thing to request when it wouldn't get you any closer to a perfect first level then just doing it the way its already done. 

For an IDEX machine, wouldnt you want to be sure that the nozzles are at the same height relative to each other without the machine compensating? I guess you could have it compensate depending on which nozzle it was using, but you wouldnt be able to print at the same time with both heads. In theory, the z height for nozzle 1 and nozzle 2 should be identical when measured at the same point over the bed, and that level should be mechanically determined by construction, not electronically. Calibrating the Z offset should be a single number, as if you only had one extruder or one z probe. Unless you only wanted to ever use one head per print...defeating the purpose of IDEX?

**Reply by huntervaneck** [reply]
Thank you for taking the time to respond in a way that I can understand. 

I think that what I'm asking for is not being conveyed as clearly as I intended. I wasn't asking for a feature that would have the user adjusting software offsets, but rather something you could print to give a visual ballpark of the current offset without changing the screws for bed leveling, or using paper to touch the nozzle. Simply a calibration to generate squares at different offsets to show you where your offset needs to be to have the best first layer spacing. This would then be changed in the firmware, as you were suggesting. 

For years I've been doing it the way you've explained, by setting the offset in the firmware and adjusting by baby stepping. If you're saying that this is the best way, that's perfectly fine. I'm thankful for the fine work you guys are doing on this slicer and I will learn from your considerations.

**Reply by supermerill** [reply]
> without changing the screws for bed leveling, or using paper to touch the nozzle

Without being the precise stuff, you have to do at least the paper stuff one time to ensure the nozzle isn't going to crash on the bed and it's not flying a mm away from the bed.
Just to nitpick
